src/openzaak/utils/middleware.py

```diff
@@ -15,3 +24,45 @@ def __call__(self, request):
 
     def log(self, request: HttpRequest):
         logger.debug("Request headers for %s: %r", request.path, request.headers)
+
+
+def get_version_mapping() -> Dict[str, str]:
+    apis = ("autorisaties", "besluiten", "catalogi", "documenten", "zaken")
+    version = settings.REST_FRAMEWORK["DEFAULT_VERSION"]
+
+    return {
+        reverse(f"api-root-{api}", kwargs={"version": version}): getattr(
+            settings, f"{api.upper()}_API_VERSION"
+        )
+        for api in apis
+    }
+
+
+class APIVersionHeaderMiddleware(_APIVersionHeaderMiddleware):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.version_mapping = get_version_mapping()
+
+    def __call__(self, request):
+        if self.get_response is None:
+            return None
+
+        response = self.get_response(request)
+
+        # not an API response, exit early
+        if not isinstance(response, Response):
+            return response
+
+        # set the header
+        version = self._get_version(request.path)
+        if version is not None:
+            response[VERSION_HEADER] = version
+
+        return response
+
+    def _get_version(self, path: str) -> Optional[str]:
+        for prefix, version in self.version_mapping.items():
+            if path.startswith(prefix):
+                return version
+        return None
```

Comment:
Feels a little crufty. You add the version to the response while we already know the version upon entering the API: In the URL-patterns.

Can't we pass to the root API URL-pattern something like: `url('/', zaken_router, kwargs={'version': settings.ZAKEN_API_VERSION})`  so that's its available to the views (for later ^_^) and in the response, we simply pass this value again without the need to look up which URL the response came from?

Reply:
but `version` is only the major versions, which is used in the URLs, while the header contains the full `major.minor.patch` version, or am I missing something?

Reply:
I meant a new version argument, not the existing one used for major version in the `v<verison>` url path, ie `request_version`=request.REQUESTED_VERSION or settings.API_VERSION

But, can be done later.